### DISCUSSION TITLE
How can I get streaming text to show up inside a field that's inside of a ListState?

### DISCUSSION THREAD

**Discussion by futzy314** [unoplatform/uno#14994] [Q&A] [answered]
Hi, I'm working on an application for chatting with AI bots. Many of them support streaming text responses, so you can see the first part of the response without waiting for the response to complete. However, I haven't been able to figure out to get that streaming response to show up in the UI. I'm using the MVUX pattern.

`Model class snippet

public IListState<Message> Messages => ListState<Message>.Value(this, () => FixupMessages(_scene.Messages));`

`XAML page snippet

    <mvux:FeedView Source="{Binding Messages}" Grid.Row="0">
      <mvux:FeedView.ValueTemplate>
        <DataTemplate>
          <ListView x:Name="SceneListView" ItemsSource="{Binding Data}" VerticalAlignment="Bottom" VerticalContentAlignment="Bottom">
            <ListView.ItemTemplate>
              <DataTemplate>
                <Border BorderBrush="White" BorderThickness="1" CornerRadius="5" Margin="5">
                  <Grid>
                    <Grid.RowDefinitions>
                      <RowDefinition Height="Auto" />
                      <RowDefinition Height="Auto" />
                    </Grid.RowDefinitions>
                    <Grid.ColumnDefinitions>
                      <ColumnDefinition Width="Auto" />
                      <ColumnDefinition Width="*" />
                    </Grid.ColumnDefinitions>

                    <TextBlock Grid.Column="0" Grid.Row="0" Text="{Binding CharacterName}" Margin="10" VerticalAlignment="Center" />
                    <Image Grid.Column="0" Grid.Row="1" Source="{Binding ImageLocation, Converter={StaticResource UrlConverter}}" Margin="10" Width="60" Height="90"/>
                    <Button Grid.Column="1" Grid.Row="0" HorizontalAlignment="Right" Width="16" Height="16" Margin="2" Style="{StaticResource IconButtonStyle}">
                      <SymbolIcon Symbol="Edit" />
                    </Button>
                    <Border Grid.Column="1" Grid.Row="1" BorderBrush="White" BorderThickness="1" Background="{Binding Color, Converter={StaticResource ColorConverter}}" CornerRadius="5" Margin="10">
                      <TextBlock behaviors:TextBlockExtensions.BindableInlines="{Binding MessageSpans}" TextWrapping="Wrap" Margin="10" />
                    </Border>
                  </Grid>
                </Border>
              </DataTemplate>
            </ListView.ItemTemplate>
          </ListView>
        </DataTemplate>
      </mvux:FeedView.ValueTemplate>
    </mvux:FeedView>`

If I'm not in streaming mode, this works as expected, I add a new message to the list in the model, it shows up in the UI. But if I try to stream it by adding an empty message to the list in the model and then updating the field bound to the TextBlock, it only shows up empty, it never updates with the text as it streams in.  I'm hoping I'm just missing something obvious. This is my first project with Uno Platform, and I'm enjoying it so far.

Thanks for any help you can provide.

**Comment by dr1rrb** [comment, marked as answer]
Hey @futzy314 👋

When working with reactive programming, entities are expected to be immutable. So here the recommendation would be to create a new instance of your entity and update (i.e. remove/add) the instance in your list:

```csharp
// Assuming that entity
public partial record Message(string Id, string Text);

private async ValueTask OnTextReceived(string messageId, string text)
{
    await Messages.UpdateAsync(message => 
    {
        var msgIndex = messages.IndexOf(msg => msg.Id == messageId);
        return msgIndex < 0 
            ? messages
            : messages.RemoveAt(index).SetItem(index, messages[index] with { Text = text }));
    }
}
```
(You can create some extension methods to reduce the noise here ;) )

If you like, you can also use the messenger of the windows community toolkit with our [ Uno.Extensions.Reactive.Messaging](https://www.nuget.org/packages/Uno.Extensions.Reactive.Messaging/) to do the same, eg:

```csharp
// Model
model_ctor(IMessenger messenger)
{
    Messages.Observe(messenger, msg => msg.Id);
}

// Business service
_messenger.Send(new EntityMessage<Message>(EntityChange.Updated, oldMsg with { Text = text }));
```

Note: In both case you should make sure to have an `Id` on your `Message` entity record and make it partial. This will allow us to generate [`IKeyEquatable` implementation](https://platform.uno/docs/articles/external/uno.extensions/doc/Learn/KeyEquality/concept.html#concept). This allows us to properly track the entities and then make sure, in the view, to notify it as an update instead of a removal and an addition, so we have proper transitions in the `ListView`.
